Add rcu-ish component building to the ractive bin - RFC

evs-chris opened the pull request:
## Description of the pull request:

This splits the ractive bin script up into modules that could conceivably be used by third parties. It stays dep-free, so there are a few things that rcu offers that aren't present here, namely source mapping and mangling various forms of dependency loading.

Instead of being very clever, this takes the dumb approach to compiling a component. It merely collects up the relevant files, converts them to the appropriate string form, does a string substitution on the script portions, and plops the result back out to STDOUT or the designated file. This leaves any module craziness up to the component author, but I'd recommend using es module format, as it translates reasonably well into other formats where needed. It's also a bit less cluttered.

Here's what this will do:

1. Read in a base template file from STDIN, a given file, or a given directory with an `index.html` or `index.ractive` and parse it as a template
2. Iterate through the top-level bits of template looking for script, template, style, and link tags
    * script tags that are `text/ractive` or `text/html` and have an `id` or `name` are turned into partials - if the script has a `src`, it is read from the designated file relative to the component file
    * script tags that are `text/javascript` (or untyped) are concatenated in order and used as the basis of the component. It's generally only recommended to have one, but this doesn't argue about having more.
    * template tags that have a `name` or `id` are parsed as partials
    * style tags have their content added to the component css string
    * link tags that are type `text/css` or `component` and rel `ractive` or `component` with an `href` will have the file at `href` read and appended to the component css string
    * any processed tags are removed from the template
3. The extra partials are parsed and have their template set to the appropriate partial id/name and their expressions added to the main template expressions
4. The css has its whitespace collapsed and it is JSON.stringify-ed
5. The script has any `$TEMPLATE` and `$CSS` identifiers replaced with the stringified template and css respectively

The whole chunk of relevant code ended up being ~120 lines, which I think is acceptable for a component compiler. External deps _should_ be as simple (when using ES module format) as `import Ractive from 'ractive'; import OtherComponent from '../OtherComponent`;` or whatever, and the result can be shared with the world by `export`ing it. I believe module bundlers like rollup would work nicely with this sort of format.

## Fixes the following issues:
#1051 #2786 and it could fix #1892 by allowing a `$PARTIALS` sub that spit out a stringified object with all of the partials, but I'm not sure how useful that is in the context of csp and more powerful yielding.

## Is breaking:
Nope.

## Reviewers:
@fskreuz and anyone else interested in having a Ractive CLI for (hopefully) less painful component building and template pre-parsing.

evs-chris committed:
add rcu-ish component building to the ractive bin

ghost commented:
Reading lodash earlier to see how it split up its functions into separate modules while keeping one with everything.

We can have Ractive loaded as is, like how we do:

```
// The regular lib gets returned
var Ractive = require('ractive'); 
```

But then have rcu functions be loaded like this. They're separate modules from the main Ractive lib, kinda like the way how lodash has one gigantic lodash.js, and have multiple independent modules for its functions.

```
var parse = require('ractive/parse');
var make = require('ractive/make');
var generateSourceMap = require('ractive/generateSourceMap');
// and so on...

// Use like rcu
```

Then the bin just becomes a CLI wrapper for those same functionality, except it just takes in options from the terminal instead of a node script.

```
// stdin input, stdout output
ractive make input.html > output.html

// file input, file output
ractive make -i input.html -o output.html

// All the rest
ractive parse ...
ractive sourcemap ...
// and so on
```

That way, it can both serve those who prefer the CLI way of doing things and the ones who make Ractive part of their node tools. Two birds with one stone n_n)b

evs-chris commented:
I think I'm only one step off of a reasonably consumable API stuck right in the ractive module then. I just need to move the output handling up a layer and let the component module work directly with strings. It will still have to load files in some way. I suppose the polite way to do that would be to accept a loading function like rcu and promisify everything. Then the bin wrapper can provide a default node file loader that handles adjusting paths to be relative to the input.

I don't think this will ever quite _replace_ rcu, as it doesn't quite have the same intentions. It looks like rcu is mostly isomorphic, but I don't see a whole lot of point to that for bundling with the ractive module. There's just so much more mileage you can get out of pre-building and bundling app and library code. I keep hearing that http2 makes having multiple requests actually preferable to bundling, but it seems like the added complexity of having a server smart enough to know which scripts it needs to push out with certain requests would be trounced by simply bundling the scripts up as a build step. I say simply because rollup is _relatively_ simple :smile:. I still haven't muddled through a webpack or browserify build.

I also don't think sourcemap support here is particularly useful. The only _actual_ source that ends up in the output is the script source, which is recommended to come from a single script tag. Everything else, being styles and template chunks, is neither executable nor sourcemap-able. And generating/interacting with the various module formats is way beyond the scope of what I think this should handle.

So, I think that leaves something like `require('ractive/lib/component')` handing back an object with a function that takes a few options (the primary string template, a file loader function, parser options) and returns a promise of a string. The the much simpler `require('ractive/lib/parse')` would hand back an object with a parse function, which takes a string template and hands back a string of compiled template, and a stringify function, which takes a compiled template and stringify-s it. Does that sound reasonable?

I think adjusting the modules like that will make this reasonably testable, which is something I was trying to figure out how best to approach.

ghost commented:
Oh snap :facepalm: Your second paragraph reminded me that rcu is more than just a node module/build time dependency. It's actually an embeddable script. I was looking at everything from a node build tool dev dependency perspective. They might have to stay separate after all. D:

fskreuz commented on bin/lib/component.js:
```diff
@@ -0,0 +1,166 @@
+      -p, --static-triple {open} {close} - set the static triple delimiters (default [[[ ]]])
+`;
+
+function build ( Ractive, string, output, opts, base ) {
```
I'm curious why `Ractive` has to be handed over via args. I believe you can easily do the following since Node does seem to cache modules.

```
var Ractive = require('../../ractive');
```

evs-chris replied:
I was just being lazy since ractive may exist at `../ractive.js` or `../build/ractive.js` depending on whether you're running from a git checkout or the actual built module.

fskreuz replied:
I see. Oki doki. 😄

evs-chris committed:
move bin/lib to lib, promisify lib functions and generally make them suitable for external consumption, and compact parsed template output a bit

evs-chris commented:
I've adjusted the library functions so that they could conceivably be used by third parties:

* `require('ractive/lib/parse').parse('some template', { parse: options }).then(str => 'do something with the stringified parsed template');`
* `require('ractive/lib/component').build('some template', { parse: options }, function readFile(path) { return aPromiseForTheFileContent; }).then(str => 'do something with the component js');`

I also added the $PARTIALS replacement, which just subs the JSON.stringify-ed partials from script tags, and adjusted the stringified template output so that it doesn't quote stuff that doesn't really need to be quoted. That saves a fair amount of space for parsed templates.

evs-chris committed:
allow the input file for the bin to specified as the last arg without a flag

evs-chris committed:
add the bin to the changelog
